language-support/ts/codegen/src/TsCodeGenMain.hs

```diff
@@ -47,36 +47,38 @@ optionsParserInfo = info (optionsParser <**> helper)
     <> progDesc "Generate TypeScript bindings from a DAR"
     )
 
-readPackages :: [FilePath] -> IO [(PackageId, (Package, Maybe String))]
+readPackages :: [FilePath] -> IO [(PackageId, (Package, Maybe PackageName))]
 readPackages dars = concatMapM readPackage dars
   where
+    readPackage :: FilePath -> IO [(PackageId, (Package, Maybe PackageName))]
     readPackage dar = do
       dar <- B.readFile dar
       let archive = Zip.toArchive $ BSL.fromStrict dar
       dalfs <- either fail pure $ DAR.readDalfs archive
       DAR.DalfManifest{packageName} <- either fail pure $ DAR.readDalfManifest archive
+      packageName <- pure (PackageName . T.pack <$> packageName)
       forM ((DAR.mainDalf dalfs, packageName) : map (, Nothing) (DAR.dalfs dalfs)) $
         \(dalf, mbPkgName) -> do
           (pkgId, pkg) <- either (fail . show)  pure $ Archive.decodeArchive Archive.DecodeAsMain (BSL.toStrict dalf)
           pure (pkgId, (pkg, mbPkgName))
 
-mergePackageMap :: [(PackageId, (Package, Maybe String))] -> Either String (Map.Map PackageId (Maybe String, Package))
+mergePackageMap :: [(PackageId, (Package, Maybe PackageName))] -> Either String (Map.Map PackageId (Maybe PackageName, Package))
```

Comment:
Might be slightly simpler if you go for `Either Text` and then do the conversion at the call site (or just use `Data.Text.IO.putStrLn`).

Reply:
Ever so slightly. Did that.